File: app-backend/src/main/java/com/fhdwapp/appbackend/service/RatingService.java
package com.fhdwapp.appbackend.service;

import com.fhdwapp.appbackend.model.Rating;

import java.util.Collection;
import java.util.Optional;

/**
 * The interface Rating service.
 */
public interface RatingService {
    /**
     * Gets all.
     *
     * @return the all
     */
    Collection<Rating> getAll();

    /**
     * Gets all by user id.
     *
     * @param userId the user id
     * @return the all by user id
     */
    Collection<Rating> getAllByUserId(Long userId);

    /**
     * Gets all by bouncy house id.
     *
     * @param bouncyHouseId the bouncy house id
     * @return the all by bouncy house id
     */
    Collection<Rating> getAllByBouncyHouseId(Long bouncyHouseId);

    /**
     * Gets by rented id.
     *
     * @param rentedId the rented id
     * @return the by rented id
     */
    Optional<Rating> getByRentedId(Long rentedId);

    /**
     * Get optional.
     *
     * @param id the id
     * @return the optional
     */
    Optional<Rating> get(Long id);

    /**
     * Create rating.
     *
     * @param rating the rating
     * @return the rating
     */
    Rating create(Rating rating);

    /**
     * Update rating.
     *
     * @param rating the rating
     * @return the rating
     */
    Rating update(Rating rating);

    /**
     * Delete boolean.
     *
     * @param id the id
     * @return the boolean
     */
    Boolean delete(Long id);

}

File: app-frontend/src/app/data-access/services/rating.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { catchError, Observable, tap, throwError } from 'rxjs';
import { CustomResponse } from '../../shared/interfaces/custom-response';
import { Rented } from '../../shared/models/rented.model';
import { Rating } from '../../shared/models/rating.model';

@Injectable({
  providedIn: 'root',
})
export class RatingService {
  constructor(private http: HttpClient) {}

  private readonly apiUrl = 'http://localhost:8080';

  /**
   * Stores ratings by user from the api in an observable
   * @param userId
   */
  ratingsByUser$ = (userId: number) =>
    <Observable<CustomResponse>>(
      this.http
        .get<CustomResponse>(`${this.apiUrl}/rating/user/${userId}`)
        .pipe(tap(console.log), catchError(this.handleError))
    );

  /**
   *
   * @param bouncyHouseId
   */
  ratingsByBouncyHouse$ = (bouncyHouseId: number) =>
    <Observable<CustomResponse>>(
      this.http
        .get<CustomResponse>(
          `${this.apiUrl}/rating/bouncy-house/${bouncyHouseId}`
        )
        .pipe(tap(console.log), catchError(this.handleError))
    );

  /**
   *
   * @param rentedId
   */
  ratingByRented$ = (rentedId: number) =>
    <Observable<CustomResponse>>(
      this.http
        .get<CustomResponse>(`${this.apiUrl}/rating/rented/${rentedId}`)
        .pipe(tap(console.log), catchError(this.handleError))
    );

  /**
   *
   * @param rating
   */
  save$ = (rating: Rating) =>
    <Observable<CustomResponse>>(
      this.http
        .post<CustomResponse>(`${this.apiUrl}/rating/save`, rating)
        .pipe(tap(console.log), catchError(this.handleError))
    );

  /**
   *
   * @param error
   * @private
   */
  private handleError(error: HttpErrorResponse): Observable<never> {
    console.log(error);
    return throwError(
      () => new Error(`An error occurred - Error code ${error.status}`)
    );
  }
}
